pubsub.go

```diff
@@ -42,71 +42,95 @@ type PubsubValueStore struct {
 
 	watchLk  sync.Mutex
 	watching map[string]*watchGroup
 
 	Validator record.Validator
 }
 
+type topicInfo struct {
+	topic *pubsub.Topic
+	evts  *pubsub.TopicEventHandler
+	sub   *pubsub.Subscription
+
+	cancel   context.CancelFunc
+	finished chan struct{}
+
+	dbWriteMx sync.Mutex
+}
+
 // KeyToTopic converts a binary record key to a pubsub topic key.
 func KeyToTopic(key string) string {
 	// Record-store keys are arbitrary binary. However, pubsub requires UTF-8 string topic IDs.
 	// Encodes to "/record/base64url(key)"
 	return "/record/" + base64.RawURLEncoding.EncodeToString([]byte(key))
 }
 
-// NewPubsubPublisher constructs a new Publisher that publishes IPNS records through pubsub.
-// The constructor interface is complicated by the need to bootstrap the pubsub topic.
-// This could be greatly simplified if the pubsub implementation handled bootstrap itself
-func NewPubsubValueStore(ctx context.Context, host host.Host, cr routing.ContentRouting, ps *pubsub.PubSub, validator record.Validator) *PubsubValueStore {
+type PubsubValueStoreOpt func(*PubsubValueStore) error
+
+// NewPubsubValueStore constructs a new ValueStore that gets and receives records through pubsub.
+func NewPubsubValueStore(ctx context.Context, host host.Host, ps *pubsub.PubSub, validator record.Validator, opts ...PubsubValueStoreOpt) (*PubsubValueStore, error) {
 	psValueStore := &PubsubValueStore{
 		ctx: ctx,
 
-		cr: cr, // needed for pubsub bootstrap
-
 		ds:                      dssync.MutexWrap(ds.NewMapDatastore()),
 		ps:                      ps,
 		host:                    host,
 		rebroadcastInitialDelay: 100 * time.Millisecond,
 		rebroadcastInterval:     time.Minute * 10,
 
-		subs:     make(map[string]*pubsub.Subscription),
-		cancels:  make(map[string]context.CancelFunc),
+		topics:   make(map[string]*topicInfo),
 		watching: make(map[string]*watchGroup),
 
 		Validator: validator,
 	}
 
+	for _, opt := range opts {
+		err := opt(psValueStore)
+		if err != nil {
+			return nil, err
+		}
+	}
+
 	psValueStore.fetch = newFetchProtocol(ctx, host, psValueStore.getLocal)
 
 	go psValueStore.rebroadcast(ctx)
 
-	return psValueStore
+	return psValueStore, nil
 }
 
-// Publish publishes an IPNS record through pubsub with default TTL
+// PutValue publishes a record through pubsub
 func (p *PubsubValueStore) PutValue(ctx context.Context, key string, value []byte, opts ...routing.Option) error {
-	// Record-store keys are arbitrary binary. However, pubsub requires UTF-8 string topic IDs.
-	// Encode to "/record/base64url(key)"
-	topic := KeyToTopic(key)
-
 	if err := p.Subscribe(key); err != nil {
 		return err
 	}
 
 	log.Debugf("PubsubPublish: publish value for key", key)
 
+	p.mx.Lock()
+	ti, ok := p.topics[key]
+	if !ok {
+		p.mx.Unlock()
+		return errors.New("could not find topic handle")
+	}
+	p.mx.Unlock()
+
+	ti.dbWriteMx.Lock()
+	defer ti.dbWriteMx.Unlock()
+	ok, err := p.putLocal(ti, key, value)
+	if err != nil {
+		return err
+	}
+	if !ok {
+		return nil
```

Comment:
Is `return nil` better than something like `return errors.New("put record that is worse than the current state")`?

Reply:
This is wrong. We need to publish regardless because both the DHT and this package will fight over updating the datastore.

Reply:
We should check to see if the new value is _at least_ as new as the local value.

Reply:
> We should check to see if the new value is at least as new as the local value.

putLocal checks to see if the new value is **better** (for IPNS ~= newer) than the local value.

> We need to publish regardless because both the DHT and this package will fight over updating the datastore.

I'm confused by what you mean here, AFAIK here are the records used by IPNS:

- Namesys: `/ipns/base32(binaryPeerID)`
- DHT: `base32(/ipns/binaryPeerID)`
- PubSub Router (here): `/base32(/ipns/binaryPeerID)`
   - Up until this PR was only storing the data in a local map datastore, so there was no "fighting"
   - This PR adds an option to use another datastore (e.g. the main one) so that we can eventually move to having a shared datastore (in order to enable third party republishing)
        - I can just remove the datastore option for now and then no one will possibly mess this up. We can have a separate PR that adds the datastore option and changes the record to look like the Namesys record.

Reply:
Got it. I assumed the datastore was shared.

Let's make the check "at least as good as" anyways. That's what we do in the DHT and, as you say, it's what we'll need to do if we share the datastore.

Reply:
> Let's make the check "at least as good as" anyways.

I'm confused, I already do that with the boolean returned by `putLocal` anyway. No?

Reply:
We're currently checking if the new value is strictly better than the existing value. We should be checking that the new value is at least as good as the new value. That way:

1. The user can explicitly re-publish an existing value.
2. If/when we start sharing a datastore with other routers, we won't see the newer value from the DHT and then decide to _not_ broadcast our value over pubsub.

Reply:
Ah, thanks for clarifying. Not an unreasonable request. lmk if it's good enough now.